osnap/analytics/analytics.py

```diff
@@ -146,15 +140,13 @@ def cluster_spatial(dataset,
 
     def _build_data(data, tracts, year, weights_type):
         df = data.loc[data.year == year].copy()
-        tracts = tracts.copy()[tracts.index.isin(df.index)]
+        tracts = tracts.copy()[tracts.geoid.isin(df.index)]
         weights = {"queen": Queen, "rook": Rook}
-        w = weights[weights_type].from_dataframe(
-            tracts.reset_index(), idVariable="geoid")
+        w = weights[weights_type].from_dataframe(tracts, idVariable="geoid")
         # drop islands from dataset and rebuild weights
         df.drop(index=w.islands, inplace=True)
```

Comment:
Do we have to drop islands? libpysal has methods to attach them?

Reply:
I believe the idea was to drop disconnected singletons, which seem to be a nuisance, but retain contiguous groups that are disconnected from the main body. We use KNN weights to attach "real" islands in https://github.com/spatialucr/osnap/blob/d8cb6235f09165fb280222bf19fd37e761292109/osnap/analytics/analytics.py#L173

In short, it's ok if the graph isn't fully connected, but not if the disconnection is induced by a single tract